### PR TITLE
Implement IPv6 support in the inet extension.

### PR DESCRIPTION
**Pull request by troycurtisjr**
I've started to use the duckdb and have several cases where the inet module has been really nice. However, it doesn't currently support IPv6, so this PR aims to add that support. There is also a somewhat related discussion about [using uhugeint to store IPv6](https://github.com/duckdb/duckdb/discussions/2945). 

The CI runs for the latest rebase are [in progress](https://github.com/troycurtisjr/duckdb/actions), but there was already a  mostly successful back before my latest rebase. There were a few failures in the [NightlyTests action](https://github.com/troycurtisjr/duckdb/actions/runs/7963746890), but nothing seemed related to the inet extension. I'm hoping the new run will have pulled in fixes from `main`.

Corresponding doc PR: [duckdb-web](https://github.com/duckdb/duckdb-web/pull/2314)

### CODE REVIEW

File: extension/inet/inet_functions.cpp [hunk 1]
```diff
@@ -3,13 +3,65 @@
...
+// the compatible representation.
 using INET_TYPE = StructTypeTernary<uint8_t, hugeint_t, uint16_t>;
 
+static uhugeint_t from_compat_addr(hugeint_t compat_addr, IPAddressType addr_type) {
+	uhugeint_t retval = static_cast<uhugeint_t>(compat_addr);
+	// Only flip the bit for order on IPv6 addresses. It can never be set in IPv4
+	if (addr_type == IPAddressType::IP_ADDRESS_V6) {
+		// The top bit is flipped when storing as the signed hugeint so that sorting
+		// works correctly. Flip it back here to have a proper unsigned value.
+		return retval ^ (uhugeint_t(1) << 127);
```
Comment (Mytherin) [inline extension/inet/inet_functions.cpp:48]: Bitwise ops on `uhugeint` are somewhat expensive - perhaps we can do the bitwise operation directly on the component that is affected instead (e.g. `new_addr.upper` or `new_addr.lower`)?

File: extension/inet/inet_functions.cpp [hunk 2]
```diff
@@ -3,13 +3,65 @@
...
 namespace duckdb {
 
+// While the address field is better represented as a uhugeint_t, the original
+// implementation used hugeint_t, so to maintain backward-compatibility it will
+// continue to be stored as signed. However, operations on the address values
+// will use the unsigned variant, so use the functions below to convert to/from
+// the compatible representation.
 using INET_TYPE = StructTypeTernary<uint8_t, hugeint_t, uint16_t>;
 
+static uhugeint_t from_compat_addr(hugeint_t compat_addr, IPAddressType addr_type) {
```
Comment (Mytherin) [inline extension/inet/inet_functions.cpp:42]: nit: Could we use the naming conventions we use elsewhere in the codebase for functions?

File: extension/inet/inet_functions.cpp [hunk 3]
```diff
@@ -3,13 +3,65 @@
...
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-function"
+#elif defined(__GNUC__)
+#if (__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
+#pragma GCC diagnostic push
+#endif
+#pragma GCC diagnostic ignored "-Wunused-function"
+#endif
+
+#include "duckdb/common/operator/numeric_cast.hpp"
```
Comment (Mytherin) [inline extension/inet/inet_functions.cpp:23]: Is this include here necessary at all? Can we perhaps include `duckdb/common/operator/cast_operators.hpp` instead? Then we might also be able to get rid of all of the pragmas.

File: test/sql/inet/test_inet_functions.test [hunk 4]
```diff
@@ -45,3 +50,8 @@ SELECT INET '127.0.0.31' - -32;
 statement error
 SELECT INET '0.0.0.0' - 32;
 ----
+
+# overflow
+statement error
+select INET '255.255.255.255' + 1
+----
```
Comment (Mytherin) [inline test/sql/inet/test_inet_functions.test:57]: Could you add an expected error message here?

File: test/sql/inet/test_ipv6_inet_type.test [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+query II
+SELECT '::ffff:0:192.0.2.128/96'::INET, '::ffff:0:c000:280/96'::INET;
+----
+::ffff:0:192.0.2.128/96
+::ffff:0:192.0.2.128/96
+
+# quibble with too many digits
+statement error
+SELECT '2001:db8:100:0:c59b:cef3:a0d189:6de4'::INET;
+----
```
Comment (Mytherin) [inline test/sql/inet/test_ipv6_inet_type.test:113]: Could we add expected errors here?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great, well done! 

Some minor comments below. [Postgres also supports the `family` function](https://www.postgresql.org/docs/current/functions-net.html) which returns whether or not a value is IPv4 or IPv6 - perhaps it makes sense to add support for that as well?

File: extension/inet/inet_functions.cpp (hunk 2 above)
Comment (troycurtisjr) [reply extension/inet/inet_functions.cpp]: Right, just snake cased it out of habit. Updated.

File: extension/inet/inet_functions.cpp (hunk 3 above)
Comment (troycurtisjr) [reply extension/inet/inet_functions.cpp]: Turns out it isn't necessary, uhugeint has a built in cast operator from hugeint, and going from always positive hugeint to uhugeint has no risk of overflow, so the lack of checks isn't a concern.

File: test/sql/inet/test_inet_functions.test (hunk 4 above)
Comment (troycurtisjr) [reply test/sql/inet/test_inet_functions.test:57]: Makes sense, I'll work on adding these cases.

**Comment by troycurtisjr** [comment]
> Thanks for the PR! Looks great, well done!
> 
> Some minor comments below. [Postgres also supports the `family` function](https://www.postgresql.org/docs/current/functions-net.html) which returns whether or not a value is IPv4 or IPv6 - perhaps it makes sense to add support for that as well?

I have a vague idea of implementing many of the helpers from postgres inet features actually, but wanted to keep this PR focused on the basic support. Still this should be an easy add and provides a programatic way to detect v4 versus v6 now that both are supported, so it seems sensible. I'll get it added in.

File: extension/inet/inet_functions.cpp (hunk 1 above)
Comment (troycurtisjr) [reply extension/inet/inet_functions.cpp]: I've updated the implementation to perform bit ops directly on the two halves of the hugeint object.

File: test/sql/inet/test_inet_functions.test (hunk 4 above)
Comment (troycurtisjr) [reply test/sql/inet/test_inet_functions.test:61]: Added expected outputs to various error tests.

**Comment by troycurtisjr** [comment]
I implemented the review comments, including adding the `family()` function. It is ready for another round of review.

**Comment by carlopi** [comment]
Single failure is on `Checks extension entries` since `family` function needs to be declared also there.
I recently made some improvement there, I would say merging (or rebasing) with main and adding this line:
```
     {"drop_fts_index", "fts", CatalogType::PRAGMA_FUNCTION_ENTRY},
     {"dsdgen", "tpcds", CatalogType::TABLE_FUNCTION_ENTRY},
     {"excel_text", "excel", CatalogType::SCALAR_FUNCTION_ENTRY},
+    {"family", "inet", CatalogType::SCALAR_FUNCTION_ENTRY},
     {"from_json", "json", CatalogType::SCALAR_FUNCTION_ENTRY},
     {"from_json_strict", "json", CatalogType::SCALAR_FUNCTION_ENTRY},
     {"from_substrait", "substrait", CatalogType::TABLE_FUNCTION_ENTRY},
```
to src/include/duckdb/main/extension_entries.hpp should make this work.

I am testing the changes here: https://github.com/carlopi/duckdb/actions/runs/8185666386

**Pull request merged**